internal/remotewrite/proxy.go

```diff
@@ -0,0 +1,174 @@
+package remotewrite
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/go-chi/chi/middleware"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	THANOS_ENDPOINT_NAME = "thanos-receiver"
+)
+
+type Endpoints struct {
+	Endpoints []Endpoint `yaml:"endpoints"`
+}
+
+type Endpoint struct {
+	Name string `yaml:"name"`
+	URL  string `yaml:"url"`
+	// +optional
+	TlsConfig *TlsConfig `yaml:"tlsConfig,omitempty"`
+	// +optional
+	BasicAuth *BasicAuth `yaml:"basicAuth,omitempty"`
+}
+
+type TlsConfig struct {
+	CA   string `yaml:"ca"`
+	Cert string `yaml:"cert"`
+	Key  string `yaml:"key"`
+}
+
+type BasicAuth struct {
+	User     string `yaml:"user"`
+	Password string `yaml:"password"`
+}
+
+func Proxy(write *url.URL, endpoints *Endpoints, logger log.Logger, r *prometheus.Registry) http.Handler {
+
+	requests := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name:        "http_proxy_requests_total",
+		Help:        "Counter of proxy HTTP requests.",
+		ConstLabels: prometheus.Labels{"proxy": "metricsv1-write"},
+	}, []string{"method"})
+	r.MustRegister(requests)
+
+	if endpoints == nil {
+		endpoints = &Endpoints{}
+	}
+
+	if write != nil {
+		endpoints.Endpoints = append(endpoints.Endpoints, Endpoint{
+			URL:  write.String(),
+			Name: THANOS_ENDPOINT_NAME,
+		})
+	}
+
+	remotewriteRequests := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name:        "remote_write_requests_total",
+		Help:        "Counter of remote write requests.",
+		ConstLabels: prometheus.Labels{"proxy": "metricsv1-remotewrite"},
+	}, []string{"code", "name"})
+	r.MustRegister(remotewriteRequests)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
Suggest to define a new function for the HandleFunc instead of inline function, because it is very important for the remotewrite. and comment this function that it is async method.

Reply:
done